src/Control/Monad/Free.hs

```diff
@@ -113,40 +136,96 @@ data Free f a = Pure a | Free (f (Free f a))
 deriving instance (Typeable f, Data (f (Free f a)), Data a) => Data (Free f a)
 #endif
 
-instance (Functor f, Eq1 f) => Eq1 (Free f) where
+instance (Functor f, PE.Eq1 f) => PE.Eq1 (Free f) where
   Pure a  ==# Pure b  = a == b
-  Free fa ==# Free fb = fmap Lift1 fa ==# fmap Lift1 fb
+  Free fa ==# Free fb = fmap PE.Lift1 fa PE.==# fmap PE.Lift1 fb
   _       ==# _ = False
 
+#ifdef LIFTED_FUNCTOR_CLASSES
+instance Eq1 f => Eq1 (Free f) where
+  liftEq eq (Pure a)  (Pure b)  = eq a b
```

Comment:
I suppose I'd slightly prefer the version with explicit sharing, but I don't have a strong opinion.